src/lem/mod.rs

```diff
@@ -422,9 +598,327 @@ mod tests {
+                },
+                Char => {
+                    let expr_aux: Cons = hash2(expr_in, expr_in);
+                    let expr_aux2: Cons = hash2(expr_aux, expr_aux);
+                    let expr_out: Cons = hash2(expr_aux2, expr_aux2);
+                    let cont_out_terminal: Terminal;
+                    return (expr_out, env_in, cont_out_terminal);
+                }
+            };
+        })
+        .unwrap();
+
+        constrain_test_helper(&lem, &Ptr::num(Fr::from_u64(42)), NumSlots::new((3, 0, 0)));
+    }
+
+    #[test]
+    fn test_hash2_slots_many_nested() {
+        let lem = lem!(expr_in env_in cont_in {
+            match_tag expr_in {
+                Num => {
+                    match_tag cont_in {
+                        Cons => {
+                            let expr_comm: Cons = hash2(expr_in, expr_in);
+                            let expr_comm2: Cons = hash2(expr_comm, expr_comm);
+                            let expr_comm3: Cons = hash2(expr_comm, expr_comm2);
+                            let expr_comm4: Cons = hash2(expr_comm, expr_comm3);
+                            let expr_comm5: Cons = hash2(expr_comm, expr_comm4);
+                        },
+                        Outermost => {
+                            let expr_outer: Cons = hash2(expr_in, expr_in);
+                            let expr_outer2: Cons = hash2(expr_outer, expr_outer);
+                        }
+                    };
+                    let expr_aux: Cons = hash2(expr_in, expr_in);
+                    let expr_out: Cons = hash2(expr_aux, expr_aux);
+                    let cont_out_terminal: Terminal;
+                    return (expr_out, env_in, cont_out_terminal);
+                },
+                Char => {
+                    let expr_aux: Cons = hash2(expr_in, expr_in);
+                    let expr_aux2: Cons = hash2(expr_aux, expr_aux);
+                    let expr_out: Cons = hash2(expr_aux2, expr_aux2);
+                    let cont_out_terminal: Terminal;
+                    return (expr_out, env_in, cont_out_terminal);
+                }
+            };
+        })
+        .unwrap();
+
+        constrain_test_helper(&lem, &Ptr::num(Fr::from_u64(42)), NumSlots::new((7, 0, 0)));
+    }
+
+    #[test]
+    fn test_hash2_slots_seq() {
+        let lem = lem!(expr_in env_in cont_in {
+            let expr_aux: Cons = hash2(expr_in, expr_in);
+            let expr_out: Cons = hash2(expr_aux, expr_aux);
+            let cont_out_terminal: Terminal;
+            return (expr_out, env_in, cont_out_terminal);
+        })
+        .unwrap();
+
+        constrain_test_helper(&lem, &Ptr::num(Fr::from_u64(42)), NumSlots::new((2, 0, 0)));
+    }
+
+    #[test]
+    fn test_hash3_slots_simple() {
+        let lem = lem!(expr_in env_in cont_in {
+            match_tag expr_in {
+                Num => {
+                    let expr_out: Cons = hash3(expr_in, expr_in, expr_in);
```

Comment:
I guess LEM doesn't care about Lurk semantics, and you're just using`Cons` in these tests 'because you can', even though it wouldn't make sense for real Lurk data. Is that right?